client/components/infinite-scroll/index.js

```diff
@@ -3,36 +3,118 @@
 /**
  * External dependencies
  */
-import { Component } from 'react';
+import React from 'react';
 import PropTypes from 'prop-types';
-import { throttle } from 'lodash';
+import { throttle, defer } from 'lodash';
+import afterLayoutFlush from 'lib/after-layout-flush';
 
 const SCROLL_CHECK_RATE_IN_MS = 400;
 
-export default class InfiniteScroll extends Component {
-	static propTypes = {
-		nextPageMethod: PropTypes.func.isRequired,
+const hasIntersectionObserver = typeof window !== 'undefined' && 'IntersectionObserver' in window;
+
+const propTypeDefinition = {
+	nextPageMethod: PropTypes.func.isRequired,
+};
+
+class InfiniteScrollWithIntersectionObserver extends React.Component {
+	static propTypes = propTypeDefinition;
+
+	observedElement = React.createRef();
+
+	hasScrolledPastBottom = false;
+
+	componentDidMount() {
+		if ( this.observedElement.current ) {
+			this.observer = new IntersectionObserver( this.handleIntersection, {
+				rootMargin: '100%',
+				threshold: 1.0,
+			} );
+			this.observer.observe( this.observedElement.current );
+		}
+	}
+
+	componentWillUnmount() {
+		if ( this.observer ) {
+			this.observer.disconnect();
+		}
+		clearTimeout( this.deferredPageFetch );
+	}
+
+	getNextPage() {
+		this.props.nextPageMethod( { triggeredByScroll: this.hasScrolledPastBottom } );
+	}
+
+	componentDidUpdate() {
+		if ( ! this.deferredPageFetch && ! this.hasScrolledPastBottom ) {
+			// We still need more pages, so schedule a page fetch.
+			this.deferredPageFetch = defer( () => {
+				if ( ! this.hasScrolledPastBottom ) {
+					this.getNextPage();
+				}
+				this.deferredPageFetch = null;
+			} );
+		}
+	}
+
+	handleIntersection = entries => {
+		if ( ! entries || ! entries[ 0 ] ) {
+			return;
+		}
+
+		if ( entries[ 0 ].isIntersecting ) {
+			this.getNextPage();
+		} else {
+			// The observed element is no longer in view, so future changes must
+			// be caused by scrolling.
+			this.hasScrolledPastBottom = true;
+		}
 	};
 
-	checkScrollPositionHandler = throttle(
-		() => this.checkScrollPosition( true ),
-		SCROLL_CHECK_RATE_IN_MS
-	);
+	render() {
+		return <div ref={ this.observedElement } />;
+	}
+}
+
+class InfiniteScrollWithScrollEvent extends React.Component {
+	static propTypes = propTypeDefinition;
 
 	componentDidMount() {
 		window.addEventListener( 'scroll', this.checkScrollPositionHandler );
-		this.checkScrollPosition( false );
+		this.checkScrollPositionHandler( false );
 	}
 
 	componentDidUpdate() {
-		this.checkScrollPosition( false );
+		this.checkScrollPositionHandler( false );
 	}
 
 	componentWillUnmount() {
 		window.removeEventListener( 'scroll', this.checkScrollPositionHandler );
+		this.clearPending();
+	}
+
+	checkScrollAfterLayout = () => {
+		this.pendingLayoutFlush = afterLayoutFlush( triggeredByScroll =>
```

Comment:
`pendingLayoutFlush` can be an instance property of the component, like `checkScrollPositionHandler`. No need to recreate it on every `checkScrollAfterLayout` call.

Calling the same function returned by `afterLayoutFlush` has another advantage: `afterLayoutFlush` will not schedule another call if no layout was performed since the previous call:
```js
pendingLayoutFlush(); // schedules callback
pendingLayoutFlush(); // noop, already scheduled
pendingLayoutFlush(); // same
// browser processes animation frame and does layout
checkScrollPosition(); // called only once even thought requested 3 times
```

Reply:
There's a complication in having `pendingLayoutFlush` as an instance property: unlike `throttle`, `afterLayoutFlush` generates a function that disregards any parameters, instead of passing them down to the wrapped function.

Reply:
```javascript
const throttled = throttle( 
    param => console.log( 'throttle: ', param ), 100 );

throttled( 'hi there' );

const afterLayout = afterLayoutFlush( 
    param => console.log( 'afterLayoutFlush: ', param ) );

afterLayout( 'hi there' );
```

generates:

```
afterLayoutFlush:  undefined
throttle:  hi there
```

Reply:
I added parameter passthrough to `afterLayoutFlush` and refactored the code based on your suggestions here. Thanks!